### PR TITLE
Skip zero samples when mapping initial data

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

This PR adds the functionality to skip samples containing only 0 when mapping initial data.

## Motivation and additional information

This prevents possibly expensive data mappings of samples where the user provided no data / a zero initial condition.

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I added a test to cover the proposed changes in our test suite.
* [ ] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I sticked to C++17 features.
* [x] I sticked to CMake version 3.16.3.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: src/precice/impl/DataContext.cpp [hunk 1]
```diff
@@ -122,21 +122,25 @@ int DataContext::mapData(std::optional<double> after)
...
+          mapping.map(stample.sample, outSample.values);
+        }
+        PRECICE_DEBUG("Mapped values (t={}) = {}", stample.timestamp, utils::previewRange(3, outSample.values));
       }
 
-      PRECICE_DEBUG("Mapped values (t={}) = {}", stample.timestamp, utils::previewRange(3, outSample.values));
-
       // Store data from mapping buffer in storage
       context.toData->setSampleAtTime(stample.timestamp, std::move(outSample));
       ++executedMappings;
```
Comment (fsimonis) [inline src/precice/impl/DataContext.cpp:146]: This still needs to be moved into the if

**Review by davidscn — COMMENTED** [review]
ASTE turbine tutorial (where this showed up) works. Changes look clean and sane. Having a test to check ensure this behavior seems useful (and rather trivial in the context of map-if-necessary) though.

File: src/precice/impl/DataContext.cpp (hunk 1 above)
Comment (fsimonis) [reply src/precice/impl/DataContext.cpp]: Done

**Comment by fsimonis** [comment]
I extended the `map-if-necessary` tests to ensure non-initialized data is skipped.

I also added tests for non-zero and zero initialized data mappings for both Serial and Parallel cases as well as Read and Write mappings.

File: tests/serial/map-initial-data/zero-data/SerialWrite.xml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    <read-data name="Data" mesh="MeshTwo" />
+  </participant>
+
+  <m2n:sockets acceptor="One" connector="Two" />
+
+  <coupling-scheme:serial-explicit>
+    <participants first="One" second="Two" />
+    <max-time-windows value="1" />
+    <time-window-size value="1.0" />
+    <exchange data="Data" mesh="MeshTwo" from="One" to="Two" initialize="true" substeps="false" />
```
Comment (davidscn) [inline tests/serial/map-initial-data/zero-data/SerialWrite.xml:35]: Could it be valuable to configure at least one of the `zero-data` cases using `initialize=false`. Should have the same effect, right?

**Comment by davidscn** [comment]
Other than that looks sane and also very nice testing coverage :+1:

File: tests/serial/map-initial-data/zero-data/SerialWrite.xml (hunk 2 above)
Comment (fsimonis) [reply tests/serial/map-initial-data/zero-data/SerialWrite.xml:35]: These are covered by the `map-if-necessary` cases, which always use `initialize=false` and never write data, thus producing only 0 data samples.
